PS-4556: Implement innodb redo log encryption

dutow opened the pull request:
This is based on WL#9290 in MySQL 8.0, while also adding a keyring based rotated key encryption.

The encryption setting supports two options:

* master_key, which works like the redo log encryption in MySQL 8.0
* keyring_key, which encrypts each new redo log tablespace with the latest percona_redo key from the keyring.

Related commits in upstream 8.0:

* 1a6dd57
* 8facbb0

laurynas-biveinis commented on client/mysqltest.cc:
```diff
@@ -3871,17 +3873,78 @@ void do_mkdir(struct st_command *command)
+  Remove the directory <dir_name>
+*/
+
+void do_force_rmdir(struct st_command *command, DYNAMIC_STRING *ds_dirname)
```
static

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_1.test:
```diff
@@ -0,0 +1,134 @@
+
+#Enable redo log encryption, should report error in server log, since keyring is not loaded.
+SET GLOBAL innodb_redo_log_encrypt = ON;
+--sleep 1
```
Needs proper synchronization

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_1.test:
```diff
@@ -0,0 +1,134 @@
+
+#Enable redo log encryption
+SET GLOBAL innodb_redo_log_encrypt = ON;
+--sleep 1
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+call mtr.add_suppression("You need to use --log-bin to make --binlog-format work");
+--enable_query_log
+
+--sleep 2
```
Why?

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+
+DROP TABLE t1;
+
+# Restart the server with keyring loaded
```
No custom restart sequences, use includes

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+--source include/wait_until_connected_again.inc
+--disable_reconnect
+
+--disable_warnings
```
Remove

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+SELECT * FROM t1 LIMIT 10;
+
+# Restart to confirm the encryption info can be retrieved properly.
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
```
Use include

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+
+# Restart to confirm the encryption info can be retrieved properly.
+--source include/kill_mysqld.inc
+--exec echo "restart:--early-plugin-load="keyring_file=$KEYRING_PLUGIN" --loose-keyring_file_data=$MYSQL_TMP_DIR/mysecret_keyring2 $KEYRING_PLUGIN_OPT" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
```
kill_and_restart_mysqld.inc

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_2.test:
```diff
@@ -0,0 +1,152 @@
+DROP DATABASE tde_db;
+
+--let SEARCH_FILE=$MYSQL_TMP_DIR/keyring_query_log
+let SEARCH_PATTERN= ALTER INSTANCE ROTATE INNODB MASTER KEY;
```
What's the point of this test?

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -895,15 +912,20 @@ fsp_header_fill_encryption_info(
-	ulint			i;
-#endif
 
+	if(is_boot) {
```
space after if

laurynas-biveinis commented on storage/innobase/include/fsp0fsp.h:
```diff
@@ -378,6 +378,16 @@ fsp_header_rotate_encryption(
 	byte*			encrypt_info,
 	mtr_t*			mtr);
 
+bool
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/fsp0fsp.h:
```diff
@@ -378,6 +378,16 @@ fsp_header_rotate_encryption(
+	byte*			encrypt_info,
+	bool is_boot);
+
+bool
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/log0log.h:
```diff
@@ -187,6 +189,11 @@ void
+/* Read the first log file header to get the encryption
+information if it exist.
+@return true if success */
+bool
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/log0log.h:
```diff
@@ -366,9 +373,24 @@ ulint
+@param[in]	log_block	log block
+@return TRUE if this block was encrypted */
+UNIV_INLINE
+bool
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/log0log.ic:
```diff
@@ -73,6 +73,44 @@ log_block_set_flush_bit(
+log_block_get_encrypt_bit(
+	const byte*	log_block)
+{
+	if (LOG_BLOCK_ENCRYPT_BIT_MASK
```
if (x) return true; return false; -> return x;

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -484,6 +493,50 @@ struct Encryption {
+	@param[in,out]	encrypt_info	encryption information
+	@param[in]	is_boot		if it's for bootstrap
+	@return true if success. */
+	static bool fill_encryption_info(byte*	key,
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -484,6 +493,50 @@ struct Encryption {
+	@param[in,out]	iv		iv
+	@param[in]	encryption_info	encrytion info.
+	@return true if success */
+	static bool decode_encryption_info(byte*	key,
```
MY_NODISCARD

laurynas-biveinis replied:
Comment not addressed

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -484,6 +493,50 @@ struct Encryption {
+	@param[in,out]	src_ptr		log block which need to encrypt
+	@param[in,out]	dst_ptr		destination area
+	@return true if success. */
+	bool encrypt_log_block(
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -484,6 +493,50 @@ struct Encryption {
+	@param[in,out]	dst		destination area
+	@param[in,out]	dst_len		size of the destination in bytes
+	@return buffer data, dst_len will have the length of the data */
+	byte* encrypt_log(
```
MY_NODISCARD

laurynas-biveinis replied:
Comment not addressed

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -501,6 +554,32 @@ struct Encryption {
+					copied to this page
+	@param[in,out]	dst		scratch area to use for decryption
+	@return DB_SUCCESS or error code */
+	dberr_t decrypt_log_block(
```
MY_NODISCARD

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -501,6 +554,32 @@ struct Encryption {
+	@param[in,out]	dst		scratch area to use for decryption
+	@param[in]	dst_len		size of the scratch area in bytes
+	@return DB_SUCCESS or error code */
+	dberr_t decrypt_log(
```
MY_NODISCARD

laurynas-biveinis replied:
Comment not addressed

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1072,240 @@ log_group_file_header_flush(
+		byte	iv[ENCRYPTION_KEY_LEN];
+
+		if (srv_read_only_mode) {
+			srv_redo_log_encrypt = false;
```
Is there a testcase for this?

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9069,8 +9188,8 @@ Encryption::create_master_key(byte** master_key)
-		ib::error() << "Encryption can't find master key, please check"
-				" the keyring plugin is loaded.";
+		ib::error() << "Encryption can't find master key"
+			    << ", please check the keyring plugin is loaded.";
```
Here and below why? Were the lines too long?

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -895,15 +912,20 @@ fsp_header_fill_encryption_info(
 
+	if(is_boot) {
+		master_key_id = 0;
+		master_key = static_cast<byte*>(ut_zalloc_nokey(
```
I'd rather use my_ allocator here too, so that it's always my_free in the end regardless of is_boot

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+	master key which master_key_id is 0. */
+	if (is_boot) {
+		master_key_id = 0;
+		master_key = static_cast<byte*>(ut_zalloc_nokey(
```
I'd rather use my_ allocator here too, so that it's always my_free regardless of is_boot

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+			      NULL, false);
+
+	if (elen == MY_AES_BAD_DATA) {
+		my_free(master_key);
```
Suspicious - if it's correct to never go here with is_boot, then either implement my suggestion to use the same allocator always, either DBUG_ASSERT(!is_boot)

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+	ptr += ENCRYPTION_MAGIC_SIZE;
+
+	mach_write_to_4(ptr, master_key_id);
+	ptr += sizeof(ulint);
```
ptr += sizeof(master_key_id) or ptr += 4 (because mach_write_to_*four*). I am not sure which one is better

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+		/* We ignore report error for recovery,
+		since the encryption info maybe hasn't writen
+		into datafile when the table is newly created. */
+		if (!recv_recovery_is_on()) {
```
return recv_recovery_is_on()

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+
+	/* Get master key id. */
+	m_key_id = mach_read_from_4(ptr);
+	ptr += sizeof(ulint);
```
Same comment as in the previous function

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+		if (m_key_id == 0) {
+			/* When m_key_id is 0, which means it's the
+			default master key for bootstrap. */
+			master_key = static_cast<byte*>(ut_zalloc_nokey(
```
Consider using my_ allocation

laurynas-biveinis replied:
Have you considered using my_ allocation?

laurynas-biveinis replied:
So what is your opinion about considering using my_ allocation?
(by now only change something if you  have to re-push/re-test otherwise)

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9258,11 +9626,211 @@ Encryption::is_encrypted_page(const byte* page)
+		len = static_cast<ulint>(elen);
+		ut_ad(len == main_len);
+
+		/* Copy remain bytes. */
```
remaining

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9258,11 +9626,211 @@ Encryption::is_encrypted_page(const byte* page)
+		       OS_FILE_LOG_BLOCK_SIZE
+		       - LOG_BLOCK_HDR_SIZE - len);
+
+		/* Encrypt the remain bytes. Since my_aes_encrypt
```
remaining

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9444,10 +10015,191 @@ Encryption::encrypt(
+
+		ut_ad(static_cast<ulint>(elen) == main_len);
+
+		/* Copy the remain bytes. */
```
remaining

laurynas-biveinis replied:
comment not addressed

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9444,10 +10015,191 @@ Encryption::encrypt(
+	byte*		ptr = src;
+	dberr_t		ret;
+
+	/* Do nothing if it's not a log request. */
```
assert is not a "do nothing"

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9444,10 +10015,191 @@ Encryption::encrypt(
+
+	/* Encrypt the log blocks one by one. */
+	while (ptr != src + src_len) {
+    //ut_a((m_type == NONE) == !is_encrypted_log(ptr));
```
Remove

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9444,10 +10015,191 @@ Encryption::encrypt(
+			log_block_get_hdr_no(ptr));
+		ut_print_buf_hex(std::cerr, ptr, OS_FILE_LOG_BLOCK_SIZE);
+		fprintf(stderr, "\n");
+	ut_print_buf(stderr, m_key, 32);
```
Indentation

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9496,6 +10250,8 @@ Encryption::decrypt(
 
 	fprintf(stderr, "Decrypting page:%lu.%lu len:%lu\n",
 		space_id, page_no, src_len);
+	//ut_print_buf(stderr, m_key, 32);
```
Remove

robgolebiowski commented on CMakeLists.txt:
```diff
@@ -13,6 +13,8 @@
 # along with this program; if not, write to the Free Software
 # Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
 
+ADD_DEFINITIONS(-DUNIV_ENCRYPT_DEBUG)
```
Remove before merging

robgolebiowski commented on mysql-test/suite/innodb/t/log_encrypt_1.test:
```diff
@@ -0,0 +1,134 @@
+# WL#9290 InnoDB: Support transparent tablespace data encryption for redo log
```
These tests will need to be made a bit more "abstract" so it could be run with different keyrings - keyring_file, keyring_vault and the future keyrings. Have a look at suite/innodb/table_encrypt_1.test (version for keyring_file) and at keyring_vault/tests/table_encrypt_1.test (version for keyring_vault).

robgolebiowski commented on mysql-test/suite/innodb/t/log_encrypt_1.test:
```diff
@@ -0,0 +1,134 @@
+# WL#9290 InnoDB: Support transparent tablespace data encryption for redo log
+# This test case will test basic redo log encryption support features.
+
+--source include/no_valgrind_without_big.inc
```
What I am missing in those tests are actual server crashes that will check if the innodb checkpoint has not been reached. Please have a look at test from MariaDB repo: innodb-redo-badkey.test and innodb-redo-nokey.test.

robgolebiowski commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -7380,8 +7388,6 @@ fil_set_encryption(
 		return(DB_NOT_FOUND);
 	}
 
-	ut_ad(algorithm != Encryption::NONE);
```
Why this was moved to the bottom of the function ? Seems like unnecessary change.

robgolebiowski commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -895,15 +912,20 @@ fsp_header_fill_encryption_info(
+		master_key_id = 0;
+		master_key = static_cast<byte*>(ut_zalloc_nokey(
+		                                ENCRYPTION_KEY_LEN));
+		memcpy(master_key, ENCRYPTION_DEFAULT_MASTER_KEY,
```
How does it work - in case of boot we are using predefined master key ? why ? This means we could not encrypt this at all ...

robgolebiowski commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -211,6 +211,21 @@ fsp_flags_to_dict_tf(
 }
+
+bool
+fsp_is_undo_tablespace(uint32 space_id)
```
unused function ?

robgolebiowski commented on storage/innobase/os/os0file.cc:
```diff
@@ -9245,9 +9364,258 @@ Encryption::get_master_key(ulint* master_key_id,
+		if (master_key == NULL) {
+			return(false);
+		}
+	}
```
I do not see information on key version. Please have a look into my email "Supporting different key versions in redo log". Since we are porting this from MySQL, we need to add support for key versioning.

dutow commented:
PR updated.

Major changes:
* Changed the variable from ON/OFF to MASTER_KEY/ROTATED_KEY/OFF
* Rotated key encryption uses a key named "percona_redo". Compared to mariadb, key version is per file instead of per block, and the key version is saved in the redo log file header. (This works without changing the log format; the other doesn't)
* Generalized the tests with include files
* Added more tests (tests which also tests undo log encryption in 8.0, but I removed the undo specific parts from them)
* Increased the data size in some tests (so they do detect some issues instead of passing even with a partially broken encryption)

There are a few formatting issues left / I left a numeric magic constant in the code which should be a define, and maybe a few assertions should be correct error messages instead (they would be quite complex to reach in a real use case, but still) - I'll fix those together with the next round of review comments.

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,158 @@
+DROP TABLE t1;
+
+# Restart the server with keyring loaded
+--exec echo "wait" > $MYSQLTEST_VARDIR/tmp/mysqld.1.expect
```
Use include files, never a hand-coded restart sequence

laurynas-biveinis commented on mysql-test/include/log_encrypt_1.inc:
```diff
@@ -0,0 +1,131 @@
+call mtr.add_suppression("\\[ERROR\\] InnoDB: Can't set redo log tablespace to be encrypted.");
+--enable_query_log
+
+#Enable redo log encryption, should report error in server log, since keyring is not loaded.
```
Space after #

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,158 @@
+--source include/have_innodb.inc
+--source include/not_embedded.inc
+
+#Suppress following messages from myslqd log
```
Space after #, "mysqld", but the whole comment looks redundant

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,158 @@
+
+SELECT * FROM t1 LIMIT 10;
+
+# Restart to confirm the encryption info can be retrieved properly.
```
Use include files, never a hand-coded restart sequence

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,158 @@
+INSERT INTO t1 select * from t1;
+
+# Restart to confirm the encryption info can be retrieved properly.
+--source include/kill_mysqld.inc
```
kill_and_restart_mysqld.inc

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+# InnoDB transparent encryption on redo log.
+# This test case will test basic encryption support features.
+ --source include/no_valgrind_without_big.inc
```
Extra leading space

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+--source include/have_innodb.inc
+--source include/not_embedded.inc
+--source include/have_innodb_max_16k.inc
+ #Suppress following messages from myslqd log
```
Redundant comment

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+call mtr.add_suppression("\\[Warning\\] InnoDB: Ignoring tablespace .* because it could not be opened");
+call mtr.add_suppression("\\[ERROR\\] InnoDB: Can't set redo log tablespace to be encrypted.");
+--enable_query_log
+ let $old_innodb_file_per_table = `SELECT @@innodb_file_per_table`;
```
Leading space

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+--enable_query_log
+ let $old_innodb_file_per_table = `SELECT @@innodb_file_per_table`;
+let $old_innodb_redo_log_encrypt = `SELECT @@innodb_redo_log_encrypt`;
+ --disable_warnings
```
Leading space

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,158 @@
+SET block_encryption_mode = 'aes-256-cbc';
+# Test encryption .
+--disable_warnings
+DROP DATABASE IF EXISTS tde_db;
```
Redundant statement, and probably the only one that can actually produce warnings here?

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+ let $old_innodb_file_per_table = `SELECT @@innodb_file_per_table`;
+let $old_innodb_redo_log_encrypt = `SELECT @@innodb_redo_log_encrypt`;
+ --disable_warnings
+DROP DATABASE IF EXISTS tde_db;
```
Redundant statement, and probably the only one that can actually produce warnings here?

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+ SHOW WARNINGS;
+--let $wait_condition=SELECT @@global.innodb_redo_log_encrypt = 0
+--source include/wait_condition.inc
+ CREATE TABLE tde_db.t4 (a BIGINT PRIMARY KEY, b LONGBLOB) ENGINE=InnoDB;
```
Leading space

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+SELECT a,LEFT(b,10) FROM tde_db.t4;
+DROP TABLE tde_db.t4;
+ --disable_warnings
+DROP DATABASE IF EXISTS tde_db;
```
Drop IF EXISTS, drop warning disable

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+SELECT a,LEFT(b,10) FROM tde_db.t4;
+FLUSH LOGS;
+let $restart_parameters = restart: --early-plugin-load="keyring_file=$KEYRING_PLUGIN" --loose-keyring_file_data=$MYSQL_TMP_DIR/my_key_log3 $KEYRING_PLUGIN_OPT --innodb_redo_log_encrypt=$redo_log_encrypt_mode;
+#--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $KEYRING_PLUGIN_OPT --plugin-dir=KEYRING_PLUGIN_PATH $KEYRING_PLUGIN keyring_file.so
```
No commented out code please

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+let $restart_parameters = restart: --early-plugin-load="keyring_file=$KEYRING_PLUGIN" --loose-keyring_file_data=$MYSQL_TMP_DIR/my_key_log3 $KEYRING_PLUGIN_OPT --innodb_redo_log_encrypt=$redo_log_encrypt_mode;
+#--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $KEYRING_PLUGIN_OPT --plugin-dir=KEYRING_PLUGIN_PATH $KEYRING_PLUGIN keyring_file.so
+#--source include/kill_and_restart_mysqld.inc
+--sleep 3
```
Any sleeps must be replaced with synchronization

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+#--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $KEYRING_PLUGIN_OPT --plugin-dir=KEYRING_PLUGIN_PATH $KEYRING_PLUGIN keyring_file.so
+#--source include/kill_and_restart_mysqld.inc
+--sleep 3
+--let $_server_id= `SELECT @@server_id`
```
No custom shutdown sequences

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+SELECT a,LEFT(b,10) FROM tde_db.t2;
+SELECT a,LEFT(b,10) FROM tde_db.t3;
+SELECT a,LEFT(b,10) FROM tde_db.t4;
+#UNINSTALL PLUGIN keyring_file;
```
No commented out code please

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+SELECT a,LEFT(b,10) FROM tde_db.t4;
+COMMIT;
+let $restart_parameters = restart: --early-plugin-load="keyring_file=$KEYRING_PLUGIN" --loose-keyring_file_data=$MYSQL_TMP_DIR/my_key_log3 $KEYRING_PLUGIN_OPT --innodb_redo_log_encrypt=$redo_log_encrypt_mode  ;
+#--replace_result $MYSQL_TMP_DIR MYSQL_TMP_DIR $KEYRING_PLUGIN_OPT --plugin-dir=KEYRING_PLUGIN_PATH $KEYRING_PLUGIN keyring_file.so
```
Same set of comments as above

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+SELECT a,LEFT(b,10) FROM tde_db.t2;
+SELECT a,LEFT(b,10) FROM tde_db.t3;
+SELECT a,LEFT(b,10) FROM tde_db.t4;
+ # Non priviledge user can not set innodb_redo_log_encrypt variable.
```
"Non-privileged"

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+disconnect con1;
+--echo # In connection default
+connection default;
+ UNINSTALL PLUGIN keyring_file;
```
Leading space

laurynas-biveinis commented on mysql-test/include/log_encrypt_4.inc:
```diff
@@ -0,0 +1,155 @@
+--source include/not_embedded.inc
+--source include/have_innodb_max_16k.inc
+
+#Suppress following messages from myslqd log
```
Redundant comment

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+FLUSH PRIVILEGES;
+--echo # In connection 1 - with encryptnonprivuser
+connect (con1,localhost,encryptnonprivuser,'noauth',);
+SELECT @@global.innodb_redo_log_encrypt ;
```
Spurious space before semicolon (in many places)

laurynas-biveinis commented on mysql-test/include/log_encrypt_5.inc:
```diff
@@ -0,0 +1,150 @@
+--source include/not_embedded.inc
+--source include/not_windows.inc
+
+#Suppress following messages from myslqd log
```
Redundant comment

laurynas-biveinis commented on mysql-test/include/log_encrypt_5.inc:
```diff
@@ -0,0 +1,150 @@
+
+
+--disable_warnings
+DROP DATABASE IF EXISTS tde_db;
```
Remove, then check whether warnings still need to be disabled

laurynas-biveinis commented on mysql-test/include/log_encrypt_5.inc:
```diff
@@ -0,0 +1,150 @@
+
+UNINSTALL PLUGIN keyring_file;
+
+#--remove_file $MYSQL_TMP_DIR/new_keyring_file
```
No commented out code

laurynas-biveinis commented on mysql-test/include/log_encrypt_6.inc:
```diff
@@ -0,0 +1,146 @@
+--source include/not_embedded.inc
+
+--disable_query_log
+#Suppress following messages from myslqd log
```
Remove comment

laurynas-biveinis commented on mysql-test/include/log_encrypt_6.inc:
```diff
@@ -0,0 +1,146 @@
+
+
+--disable_warnings
+DROP DATABASE IF EXISTS tde_db;
```
Remove, then check whether warnings still need disabling

laurynas-biveinis commented on mysql-test/include/restart_mysqld_no_echo.inc:
```diff
@@ -0,0 +1,46 @@
+
```
I believe we have restart_hide_parameters in restart_mysqld.inc for this? Better not to have two different means for the same thing

dutow replied:
I can change that, but I actually backported this file from 8.0

laurynas-biveinis replied:
OK then

laurynas-biveinis commented on mysql-test/include/start_mysqld_no_echo.inc:
```diff
@@ -0,0 +1,20 @@
+if (!$restart_parameters)
```
Likewise

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_kill.test:
```diff
@@ -0,0 +1,351 @@
+eval INSTALL PLUGIN keyring_file SONAME '$KEYRING_PLUGIN';
+
+--disable_warnings
+DROP TABLE IF EXISTS t1;
```
Remove

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_kill.test:
```diff
@@ -0,0 +1,351 @@
+
+--disable_warnings
+DROP TABLE IF EXISTS t1;
+DROP DATABASE IF EXISTS tde_db;
```
Remove, check if warning disable is still required then

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_kill.test:
```diff
@@ -0,0 +1,351 @@
+
+
+--connection default
+--sleep 3
```
Please cleanup (no sleep, include files for restart)

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_kill.test:
```diff
@@ -0,0 +1,351 @@
+send call tde_db.query_table('tde_db.t_non_encrypt');
+
+--connection default
+--sleep 20
```
Replace with sync of some sort

laurynas-biveinis commented on mysql-test/suite/innodb/t/log_encrypt_kill.test:
```diff
@@ -0,0 +1,351 @@
+
+--connection default
+--sleep 20
+--echo # Starting server with keyring plugin
```
Use include file

laurynas-biveinis commented on storage/innobase/include/log0log.h:
```diff
@@ -139,6 +139,8 @@ Closes the log.
 log_close(void);
+void
+log_enable_encryption_if_set();
 /*===========*/
```
The new declaration squeezed in between log_close declaration and its decorative comment

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -485,6 +498,50 @@ struct Encryption {
+	@param[in]	iv		encryption iv
+	@param[in,out]	encrypt_info	encryption information
+	@return true if success. */
+  MY_NODISCARD
```
Check indentation

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -485,6 +498,50 @@ struct Encryption {
+	@param[in,out]	src_ptr		log block which need to encrypt
+	@param[in,out]	dst_ptr		destination area
+	@return true if success. */
+  MY_NODISCARD
```
Check indentation

laurynas-biveinis commented on storage/innobase/include/os0file.h:
```diff
@@ -502,6 +559,33 @@ struct Encryption {
+					copied to this page
+	@param[in,out]	dst		scratch area to use for decryption
+	@return DB_SUCCESS or error code */
+  MY_NODISCARD
```
Check indentation

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+bool
+log_read_encryption()
+{
+	ulint		log_space_id = SRV_LOG_SPACE_FIRST_ID;
```
Declare variables at first use, const when possible

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+		}
+
+	}
+	if(encrypted_log) {
```
Space after if

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+	if (srv_redo_log_encrypt != REDO_LOG_ENCRYPT_OFF
+	    && !FSP_FLAGS_GET_ENCRYPTION(space->flags)
+	    && strlen(server_uuid) > 0) {
+		dberr_t err;
```
Declare locals at first uses

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+			if (my_key_fetch(PERCONA_REDO_KEY_NAME, &redo_key_type, NULL,
+						reinterpret_cast<void**>(&rkey), &klen) || rkey == NULL)
+			{
+				if(my_key_generate(PERCONA_REDO_KEY_NAME, "AES", NULL, ENCRYPTION_KEY_LEN)) {
```
Space after if

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+				if(my_key_generate(PERCONA_REDO_KEY_NAME, "AES", NULL, ENCRYPTION_KEY_LEN)) {
+					DBUG_ASSERT(0);
+				}
+				if(my_key_fetch(PERCONA_REDO_KEY_NAME, &redo_key_type, NULL,
```
Space after if

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+						reinterpret_cast<void**>(&rkey), &klen) || rkey == NULL)
+			{
+				if(my_key_generate(PERCONA_REDO_KEY_NAME, "AES", NULL, ENCRYPTION_KEY_LEN)) {
+					DBUG_ASSERT(0);
```
Is it really "cannot-happen" for release build? If yes, please add a comment

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+				}
+				if(my_key_fetch(PERCONA_REDO_KEY_NAME, &redo_key_type, NULL,
+							reinterpret_cast<void**>(&rkey), &klen)) {
+					DBUG_ASSERT(0);
```
Is it really "cannot-happen" for release build? If yes, please add a comment

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -7412,11 +7420,10 @@ fil_set_encryption(
 
 	if (key == NULL) {
 		Encryption::random_value(space->encryption_key);
+		space->encryption_key_version = 65536;
```
Magic constant, please name it

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+	}
+
+	if (space->encryption_type != Encryption::NONE
+	    && space->encryption_key_version ==65536 
```
The same magic constant

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+			if (my_key_store(PERCONA_REDO_KEY_NAME, "AES", NULL,
+						space->encryption_key, ENCRYPTION_KEY_LEN))
+			{
+				ut_ad(0);
```
Cannot happen in release build?

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+			char *redo_key_type = NULL;
+			byte *rkey = NULL;
+			unsigned char *rkey2 = NULL;
+			if(my_key_fetch(PERCONA_REDO_KEY_NAME, &redo_key_type, NULL,
```
Space after if

laurynas-biveinis replied:
Still space after if, but only fix if you have to re-push/re-test otherwise

laurynas-biveinis commented on storage/innobase/log/log0log.cc:
```diff
@@ -1072,6 +1074,398 @@ log_group_file_header_flush(
+			unsigned char *rkey2 = NULL;
+			if(my_key_fetch(PERCONA_REDO_KEY_NAME, &redo_key_type, NULL,
+						reinterpret_cast<void**>(&rkey), &klen)) {
+				DBUG_ASSERT(0);
```
Cannot happen in release build?

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -2193,6 +2210,58 @@ os_file_encrypt_page(
+	ulint*			n)
+{
+
+	byte*		encrypted_log;
```
Declare locals at first use

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -7297,23 +7386,47 @@ AIO::reserve_slot(
-			os_free_block(slot->buf_block);
+			slot->buf_block = encrypted_block;
+		} else {
+			/* Skip encrypt log file header */
```
encrypted

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -7297,23 +7386,47 @@ AIO::reserve_slot(
 
-		slot->buf_block = encrypted_block;
 		slot->buf = static_cast<byte*>(src_buf);
+
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -8722,14 +8835,21 @@ os_file_set_umask(ulint umask)
 {
 	os_innodb_umask = umask;
 }
+
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -8722,14 +8835,21 @@ os_file_set_umask(ulint umask)
 #include "mach0data.h"
+#include "fsp0fsp.h"
 #include "fil0fil.h"
+
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9061,7 +9181,7 @@ Encryption::create_master_key(byte** master_key)
 	if (ret || *master_key == NULL) {
 		ib::error() << "Encryption can't find master key, please check"
-				" the keyring plugin is loaded.";
+			              " the keyring plugin is loaded.";
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9114,7 +9234,7 @@ Encryption::get_master_key(ulint master_key_id,
 		*master_key = NULL;
 		ib::error() << "Encryption can't find master key, please check"
-				" the keyring plugin is loaded.";
+			              " the keyring plugin is loaded.";
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9227,7 +9347,7 @@ Encryption::get_master_key(ulint* master_key_id,
 		*master_key = NULL;
 		ib::error() << "Encryption can't find master key, please check"
-				" the keyring plugin is loaded.";
+			              " the keyring plugin is loaded.";
```
Spurious

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9236,9 +9356,239 @@ Encryption::get_master_key(ulint* master_key_id,
+				      byte*	iv,
+				      byte*	encrypt_info)
+{
+	byte*			ptr;
```
Declare locals at first use

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9236,9 +9356,239 @@ Encryption::get_master_key(ulint* master_key_id,
+				   byte*	iv,
+				   byte*	encryption_info)
+{
+	byte*			ptr;
```
Declare locals at first use

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9249,11 +9599,211 @@ Encryption::is_encrypted_page(const byte* page)
+	byte*			src_ptr,
+	byte*			dst_ptr)
+{
+	ulint		len = 0;
```
Declare locals at first use, const when possible

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9435,10 +9988,189 @@ Encryption::encrypt(
+	byte*			src,
+	byte*			dst)
+{
+	ulint		data_len;
```
Declare locals at first use, const when possible

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9435,10 +9988,189 @@ Encryption::encrypt(
+	remain_len = data_len - main_len;
+
+	ptr += LOG_BLOCK_HDR_SIZE;
+	switch(m_type) {
```
Space after switch

laurynas-biveinis commented on storage/innobase/srv/srv0start.cc:
```diff
@@ -460,6 +460,26 @@ create_log_files(
 	ut_a(log_space != NULL);
 
+      /* Once the redo log is set to be encrypted,
++       initialize encryption information. */
```
Extra +

robgolebiowski commented:
Possible values for the variable should be : MASTER_KEY/KEYRING_KEY

satya-bodapati commented:
One suggestion about keyring loading and suppression messages in test. There is no need for suppression message (unless explicitly caused by test). 

The warnings come from non availablity of keyring plugin at startup. I have modified MTR to provide a way to 'early-load' plugin. See mysql-test/suite/innodb/t/percona_sys_tablespace_encrypt-master.opt

$KEYRING_PLUGIN_OPT
$KEYRING_PLUGIN_EARLY_LOAD
--keyring_file_data=$MYSQL_TMP_DIR/mysecret_keyring
 
You don't see a single suppression in the mysql-test/suite/innodb/t/percona_sys_tablespace_encrypt.test. This is because of KEYRING_PLUGIN_EARLY_LOAD changes to MTR framework

Another benefit of this KEYRING_PLUGIN_EARLY_LOAD is that all subsequent restarts automatically do early-plugin-load. 

I know that Oracle tests have these suppressions but we can get rid most of them with the above suggested changes.

dutow commented:
Updated.

* Fixed test / source code comments
* Changed the asserts to print out actual error messages - those should only happen if something in the keyring plugin is broken, but this way even that reports an error.
* Renamed ROTATED_KEY to KEYRING_KEY
* Removed warning supressions from most tests. I couldn't remove it from all, as some tests do generate those errors/warnings to test them.

laurynas-biveinis commented on mysql-test/include/log_encrypt_2.inc:
```diff
@@ -0,0 +1,129 @@
+let $innodb_file_per_table = `SELECT @@innodb_file_per_table`;
+
+SET GLOBAL innodb_file_per_table = 1;
+SELECT @@innodb_file_per_table;
```
Two very redundant lines, better to use assert.inc or nothing. OK to fix after the merge or never.

laurynas-biveinis commented on mysql-test/include/log_encrypt_3.inc:
```diff
@@ -0,0 +1,206 @@
+--source include/restart_mysqld_no_echo.inc
+
+SET GLOBAL innodb_redo_log_encrypt = 0;
+SELECT @@global.innodb_redo_log_encrypt;
```
Not sure what these SELECTs past SET are supposed to test (OK to fix post-merge or never)

laurynas-biveinis commented on storage/innobase/fil/fil0fil.cc:
```diff
@@ -7512,7 +7519,9 @@ fil_encryption_rotate()
 	     space != NULL; ) {
 		/* Skip unencypted tablespaces. */
-		if (srv_is_undo_tablespace(space->id)
+		/* Encrypted redo log tablespaces is handled in function
```
"tablespaces are"
Fix only if you have to re-push and re-test otherwise

laurynas-biveinis commented on storage/innobase/fsp/fsp0fsp.cc:
```diff
@@ -985,6 +1015,7 @@ fsp_header_fill_encryption_info(
 	mach_write_to_4(ptr, crc);
 
 	my_free(master_key);
+
```
Spurious diff

laurynas-biveinis commented on storage/innobase/include/log0log.h:
```diff
@@ -139,7 +139,12 @@ Closes the log.
 log_close(void);
-/*===========*/
+
+/************************************************************//**
```
/** Enables redo ... */
fix only if you have to re-push/re-test the PR otherwise

laurynas-biveinis commented on storage/innobase/include/log0log.h:
```diff
@@ -366,9 +377,26 @@ ulint
+@param[in]	val		value to set */
+UNIV_INLINE
+void
+log_block_set_encrypt_bit(
```
log_block_set_encrypt_bit(byte* log_block, bool val);
Fix only if you have to re-push/re-test otherwise

laurynas-biveinis commented on storage/innobase/os/os0file.cc:
```diff
@@ -9061,7 +9178,7 @@ Encryption::create_master_key(byte** master_key)
 	if (ret || *master_key == NULL) {
 		ib::error() << "Encryption can't find master key, please check"
-				" the keyring plugin is loaded.";
+			       " the keyring plugin is loaded.";
```
Spurious diffs, but only fix if you have to re-push

dutow commented:
I fixed the formatting only comments, as I forgot to add the rerecorded result of perfschema.show_sanity. Will 

I also found out, that the sleeps were likely there to "fix" CI issues, as without those, jenkins release jobs report test failures: https://ps.cd.percona.com/job/percona-server-5.7-param/89/
I can't reproduce this exact failure locally even with a high parallel repeated test of my new tests :( (assertion failure in a release build, uh), but I do see some failures under heavy loads locally - it just needed a lot more parallel processes than I originally used for testing. I'll try to do some more tests for this later today, but the same can happen with 8.0 & without the sleeps.

dutow commented:
I fixed the formatting only comments, as I forgot to add the rerecorded result of perfschema.show_sanity. Will 

I also found out, that the sleeps were likely there to "fix" CI issues, as without those, jenkins release jobs report test failures: https://ps.cd.percona.com/job/percona-server-5.7-param/89/
I can't reproduce this exact failure locally even with a high parallel repeated test of my new tests :( (assertion failure in a release build, uh), but I do see some failures under heavy loads locally - it just needed a lot more parallel processes than I originally used for testing. I'll try to do some more tests for this later today, but the same can happen with 8.0 & without the sleeps.

dutow commented:
I fixed the formatting only comments, as I forgot to add the rerecorded result of perfschema.show_sanity. Will 

I also found out, that the sleeps were likely there to "fix" CI issues, as without those, jenkins release jobs report test failures: https://ps.cd.percona.com/job/percona-server-5.7-param/89/
I can't reproduce this exact failure locally even with a high parallel repeated test of my new tests :( (assertion failure in a release build, uh), but I do see some failures under heavy loads locally - it just needed a lot more parallel processes than I originally used for testing. I'll try to do some more tests for this later today, but the same can happen with 8.0 & without the sleeps.

dutow commented:
I fixed the formatting only comments, as I forgot to add the rerecorded result of perfschema.show_sanity. Will 

I also found out, that the sleeps were likely there to "fix" CI issues, as without those, jenkins release jobs report test failures: https://ps.cd.percona.com/job/percona-server-5.7-param/89/
I can't reproduce this exact failure locally even with a high parallel repeated test of my new tests :( (assertion failure in a release build, uh), but I do see some failures under heavy loads locally - it just needed a lot more parallel processes than I originally used for testing. I'll try to do some more tests for this later today, but the same can happen with 8.0 & without the sleeps.

laurynas-biveinis commented:
Sleep presence should not block trunk merge in this case, although we will must fix them there afterwards

dutow commented:
Last update, with disabled tests: https://ps.cd.percona.com/job/percona-server-5.7-param/97/#showFailuresLink

No failure which occurs on every configuration, except for percona_bug1289599 which seems unrelated.
